module/VuFind/src/VuFind/Db/Service/TagService.php

```diff
@@ -71,6 +80,346 @@ public function __construct(
+            }
+        }
+    }
+
+    /**
+     * Check whether or not the specified tags are present in the table.
+     *
+     * @param array $ids IDs to check.
+     *
+     * @return array     Associative array with two keys: present and missing
+     */
+    public function checkForTags($ids)
+    {
+        // Set up return arrays:
+        $retVal = ['present' => [], 'missing' => []];
+
+        // Look up IDs in the table:
+        $dql = 'SELECT IDENTITY(rt.tag) FROM ' . $this->getEntityClass(ResourceTags::class) . ' rt '
+            . 'WHERE rt.tag IN (:ids)';
+        $query = $this->entityManager->createQuery($dql);
+        $query->setParameters(compact('ids'));
+        $results = $query->getSingleColumnResult();
+
+        // Record all IDs that are present:
+        $retVal['present'] = array_unique($results);
+
+        // Detect missing IDs:
+        $retVal['missing'] = array_diff($ids, $retVal['present']);
+
+        // Send back the results:
+        return $retVal;
+    }
+
+    /**
+     * Get resources associated with a particular tag.
+     *
+     * @param string $tag  Tag to match
+     * @param string $user ID of user owning favorite list
+     * @param string $list ID of list to retrieve (null for all favorites)
+     *
+     * @return array
+     */
+    public function getResourceIDsForTag($tag, $user, $list = null)
+    {
+        $dql = 'SELECT DISTINCT(rt.resource) AS resource_id '
+            . 'FROM ' . $this->getEntityClass(ResourceTags::class) . ' rt '
+            . 'JOIN rt.tag t '
+            . 'WHERE ' . ($this->caseSensitive ? 't.tag = :tag' : 'LOWER(t.tag) = LOWER(:tag) ')
+            . 'AND rt.user = :user ';
+
+        $user = is_object($user) ? $user : $this->entityManager->getReference(User::class, $user);
+        $parameters = compact('tag', 'user');
+        if (null !== $list) {
+            $list = is_object($list) ? $list : $this->entityManager->getReference(UserList::class, $list);
+            $dql .= 'AND rt.list = :list';
+            $parameters['list'] = $list;
+        }
+        $query = $this->entityManager->createQuery($dql);
+        $query->setParameters($parameters);
+        $result =  $query->getSingleColumnResult();
+        return $result;
+    }
+
+    /**
+     * Get lists associated with a particular tag.
+     *
+     * @param string|array      $tag        Tag to match
+     * @param null|string|array $listId     List ID to retrieve (null for all)
+     * @param bool              $publicOnly Whether to return only public lists
+     * @param bool              $andTags    Use AND operator when filtering by tag.
+     *
+     * @return array
+     */
+    public function getListsForTag(
+        $tag,
+        $listId = null,
+        $publicOnly = true,
+        $andTags = true
+    ) {
+        $tag = (array)$tag;
+        $listId = $listId ? (array)$listId : null;
+        $dql = 'SELECT IDENTITY(rt.list) as list, COUNT(DISTINCT(rt.tag)) AS tagcnt '
```

Comment:
When working on my pull request to fix existing bugs in this method, I discovered that selecting tagcnt here and using it in the HAVING clause below does not work with PostgreSQL. There is no reason to select it anyway, so I suggest removing this clause here, and changing the HAVING below to directly embed `COUNT(DISTINCT(rt.tag))` in place of the name `tagcnt`.